File: src/搜索/BFS/subject/P12258_蓝桥杯2024国JavaB_背包问题/Main.cpp
#include <iostream>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <vector>
#include <functional>
using namespace std;

typedef pair<long long, int> pii;

int main() {
    int n, k, l;
    scanf("%d %d %d", &n, &k, &l);

    // 输入并去重物品重量
    unordered_set<long long> ws;
    for (int i = 0; i < n; ++i) {
        long long x;
        scanf("%lld", &x);
        ws.insert(x);
    }
    vector<long long> unique_w(ws.begin(), ws.end());

    priority_queue<pii, vector<pii>, greater<>> pq;
    unordered_map<long long, int> max_num;
    unordered_set<long long> answer_set;

    // 初始化队列
    for (auto& w : unique_w) {
        if (!max_num.count(w) || max_num[w] < 1) {
            max_num[w] = 1;
            pq.emplace(w, 1);
        }
    }

    int count = 0;
    long long ans = -1;

    while (!pq.empty()) {
        auto [cur_wei, cur_num] = pq.top();
        pq.pop();

        // 跳过非最大num的状态
        if (max_num[cur_wei] != cur_num) continue;

        // 满足条件则加入答案集合
        if (cur_num >= k) {
            if (answer_set.insert(cur_wei).second) {
                if (++count == l) {
                    ans = cur_wei;
                    break;
                }
            }
        }

        // 生成新的状态
        for (auto& w : unique_w) {
            long long new_wei = cur_wei + w;
            int new_num = cur_num + 1;

            // 仅当新数目更大时更新
            if (!max_num.count(new_wei) || new_num > max_num[new_wei]) {
                max_num[new_wei] = new_num;
                pq.emplace(new_wei, new_num);
            }
        }
    }

    printf("%lld\n", ans);
    return 0;
}

File: src/搜索/BFS/subject/P12258_蓝桥杯2024国JavaB_背包问题/Main.java
package 搜索.BFS.subject.P12258_蓝桥杯2024国JavaB_背包问题;

import java.util.*;
import java.io.*;

public class Main {
    static class Pair {
        long wei;
        int num;
        Pair(long wei, int num) {
            this.wei = wei;
            this.num = num;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] firstLine = br.readLine().split(" ");
        int n = Integer.parseInt(firstLine[0]);
        int k = Integer.parseInt(firstLine[1]);
        int l = Integer.parseInt(firstLine[2]);

        // 输入并去重物品重量
        Set<Long> ws = new HashSet<>();
        String[] weights = br.readLine().split(" ");
        for (String s : weights) {
            ws.add(Long.parseLong(s));
        }
        List<Long> uniqueW = new ArrayList<>(ws);

        PriorityQueue<Pair> pq = new PriorityQueue<>((a, b) -> Long.compare(a.wei, b.wei));
        Map<Long, Integer> maxNum = new HashMap<>();
        Set<Long> answerSet = new HashSet<>();

        // 初始化队列
        for (Long w : uniqueW) {
            if (!maxNum.containsKey(w) || maxNum.get(w) < 1) {
                maxNum.put(w, 1);
                pq.add(new Pair(w, 1));
            }
        }

        int count = 0;
        long ans = -1;

        while (!pq.isEmpty()) {
            Pair cur = pq.poll();
            long curWei = cur.wei;
            int curNum = cur.num;

            // 跳过非最大num的状态
            if (maxNum.getOrDefault(curWei, 0) != curNum) continue;

            // 满足条件则加入答案集合
            if (curNum >= k) {
                if (answerSet.add(curWei)) {
                    if (++count == l) {
                        ans = curWei;
                        break;
                    }
                }
            }

            // 生成新的状态
            for (Long w : uniqueW) {
                long newWei = curWei + w;
                int newNum = curNum + 1;

                // 仅当新数目更大时更新
                if (newNum > maxNum.getOrDefault(newWei, 0)) {
                    maxNum.put(newWei, newNum);
                    pq.add(new Pair(newWei, newNum));
                }
            }
        }

        System.out.println(ans);
    }
}